trie/database.go

```diff
@@ -85,22 +86,109 @@ type Database struct {
 // cachedNode is all the information we know about a single cached node in the
 // memory database write layer.
 type cachedNode struct {
-	blob     []byte              // Cached data block of the trie node
+	node node   // Cached collapsed trie node, excludes raw
+	raw  []byte // Cached raw data (code), excludes node
+	size uint16 // Byte size of the useful cached data
+
 	parents  int                 // Number of live nodes referencing this one
 	children map[common.Hash]int // Children referenced by this nodes
 
 	flushPrev common.Hash // Previous node in the flush-list
 	flushNext common.Hash // Next node in the flush-list
 }
 
+// rlp returns the raw rlp encoded blob of the cached node, either directly from
+// the cache, or by regenerating it from the collapsed node.
+func (n *cachedNode) rlp() []byte {
+	if n.raw != nil {
+		return n.raw
+	}
+	blob, err := rlp.EncodeToBytes(n.node)
+	if err != nil {
+		panic(err)
+	}
+	return blob
+}
+
+// childs returns all the tracked children of this node, both the implicit ones
+// from inside the node as well as the explicit ones from outside the node.
+func (n *cachedNode) childs() []common.Hash {
+	children := make([]common.Hash, 0, 16)
+	for child := range n.children {
+		children = append(children, child)
+	}
+	if n.node != nil {
+		gatherChildren(n.node, &children)
+	}
+	return children
+}
+
+// gatherChildren traverses the node hierarchy of a collapsed storage node and
+// retrieves all the hashnode children.
+func gatherChildren(n node, children *[]common.Hash) {
+	switch n := n.(type) {
+	case *shortNode:
+		gatherChildren(n.Val, children)
+
+	case *fullNode:
+		for i := 0; i < 16; i++ {
+			gatherChildren(n.Children[i], children)
+		}
+	case hashNode:
+		*children = append(*children, common.BytesToHash(n))
+	}
+}
+
+// expandNode traverses the node hierarchy of a collapsed storage node and converts
+// all fields and keys into expanded memory form.
+func expandNode(n node, cachgen uint16) node {
+	switch n := n.(type) {
+	case *shortNode:
+		// Short nodes need key and child expansion
+		return &shortNode{
+			Key: compactToHex(n.Key),
+			Val: expandNode(n.Val, cachgen),
+			flags: nodeFlag{
+				hash: n.flags.hash,
+				gen:  cachgen,
+			},
+		}
+
+	case *fullNode:
+		// Full nodes need child expansion
+		node := &fullNode{
+			flags: nodeFlag{
+				hash: n.flags.hash,
+				gen:  cachgen,
+			},
+		}
+		for i := 0; i < 17; i++ { // expand the term too (17)
+			node.Children[i] = expandNode(n.Children[i], cachgen)
+		}
+		return node
+
+	case valueNode:
+		// Values need to convert from "" to nil
+		if n == nil {
```

Comment:
This is a noop, `return n` should be enough.

Reply:
It is not a noop, `n` is `valueNode(nil)`, which is different from `nil` (the former can be case to a `valueNode`, the latter cannot. Yay Go quirks.

PS: replacing this with `return n` will panic the trie tests.